### DISCUSSION TITLE
remote.configure error while trying to configure s3

### DISCUSSION THREAD

**Discussion by JithenJ** [seaweedfs/seaweedfs#4283] [Q&A]
I am trying to run Seaweed FS to store files in S3, but when i try to execute remote.configure I get the below error.Steps I followed are
1. started master  `weed master -mdir=/path/to/directory`
2. started volume `weed volume -dir=/path/to/directory -mserver=localhost:9333`
3. started the filer `weed filer -master=localhost:9333`
4. Now I switched to weed shell to try and configure a cloud storage for my volumes

> `remote.configure -name=cloud1 -s3.access_key=xxx -s3.secret_key=xxx`
error: getOrCreateConnection : fail to dial : failed to build resolver: passthrough: received empty target in Build()

I do have access to a specific bucket in S3 since I can connect it using the aws cli command.

Any help is much appreciated

**Comment by chrislusf** [comment]
See `remote.configure -h`. Here are the options:
```
  -s3.access_key string
    	s3 access key
  -s3.endpoint string
    	endpoint for s3-compatible local object store
  -s3.force_path_style
    	s3 force path style (default true)
  -s3.region string
    	s3 region (default "us-east-2")
  -s3.secret_key string
    	s3 secret key
  -s3.storage_class string
    	s3 storage class
  -s3.v4_signature
    	s3 V4 signature
 ```

Very likely you are missing endpoint.

**Reply by JithenJ** [reply]
@chrislusf Thanks for the prompt reply!

In fact I had given an endpoint. But after revisiting the IAM policies to connect to that bucket I was able to successfully connect with it. 

I had another Question if I may, my use case for using Seaweed is that I would like to store some files which can be easily accessed on demand, the reason I wanted to use use AWS S3 was that I did not have to worry much about scaling the file system.

My tests are on localhost and I have mounted one of the AWS S3 buckets to my local directory. 

Steps I did to achieve this is as below
1. Started the master/ volume and Filer. My Filer is backed by a postgresDB, I have configured that in the _filer.toml_ file
`weed server -s3 -filer=true -dir=/path/to/local/directory`

2. Configure the S3 access and specify the bucket you want to read/ write to
`weed shell`
`s3.configure -user adminuser -access_key=xxx -secret_key=xxx -buckets=<aws_s3_bucketname> -actions=Read,Write,List,Tagging,Adminf`

3. Configure the remote storage
`remote.configure -name=s5 -type=s3 -s3.access_key=xxx -s3.secret_key=xxx -s3.region=eu-west-1`

4. Mount the remote storage to a local directory
`remote.mount -dir=/path/to/mount/cloud/data/directory -remote=s5/<aws_s3_bucketname> -nonempty`

5. After the above steps I can see that the Filer has synced with the files in AWS s3 bucket. I could see that the filer has got meta data of the 2 files that I had in the AWS S3 bucket.

6. My requirement is two way, I would like to read the files from AWS s3 bucket as well as push files from local into that bucket. Hence I just uploaded few files into the `/path/to/mount/cloud/data/directory`, and then ran the below command thinking that it will push all these files into aws S3. But it did not... 
`weed filer.remote.sync -dir=/path/to/mount/cloud/data/directory`

Not sure what I am missing here. Any help is much appreciated.

**Reply by chrislusf** [reply]
try to use an earlier `-timeAgo` parameter value in `weed filer.remote.sync -dir=/path/to/mount/cloud/data/directory`

**Reply by JithenJ** [reply]
Thanks @chrislusf 

I tried with the `-timeAgo=5h` , although it showed that it will sync from past 5 hours, I could not find my files being copied to the remote AWS S3 bucket. 

I also ended up having another issue, I was trying to test my setup, and in the process I cleared my postgresDB that was my filer store for my filer. And ever since I have been getting the below error. 
`error: find configuration for : filer: no entry is found in filer store`

I tried to delete all my dir and restarted weed master/vol/filer
I even changed my database to point to a completely new one. Still no luck.

Would you kindly suggest me where I might be going wrong in the above 2 issues please?

**Reply by chrislusf** [reply]
the directories and files meta information are in your postgresDB. If you clear it, the meta data is gone.

you can see the logs with `weed -v=4 filer.remote.sync ...`

**Reply by JithenJ** [reply]
Thanks for the quick response @chrislusf 

I have deleted all the dir and started with a new dir. And also changed the filer to point to a completely new DB and a new s3 bucket. 
Was wondering why it is still expecting the meta data ? I was starting from fresh again.

Lesson learnt, I will not try clean wipe again. I just assumed that if it points to a new DB store, s3 bucket, it would not be a problem.

After running the command, this is what I see: 

`I0309 22:38:50.078858 config.go:46 Reading : Config File "security" Not Found in "[/Users/<username> /Users/<username>/.seaweedfs /usr/local/etc/seaweedfs /etc/seaweedfs]"`

Not sure what is expecting here.
I dont mind loosing my data, as I am still doing a POC, but I need my filer to start up and be able to mount to an S3 bucket. 

Right now my filer is not starting up at all. I tried with a basic filestore as leveldb as well.
But no luck :(

Any help is much appreciated please.

**Reply by JithenJ** [reply]
@chrislusf 
I have had some progress now. 
Thanks to your log command I was able to understand what the actual problem was.

I sorted by filer problem +  I have also created the `security.toml` file that it was complaining about. 

Now when I run the command 
`weed filer.remote.sync -filer=localhost:8888 -dir=/path/to/mount/cloud/data/directory -timeAgo=7h`

I can see that it is trying to sync
```
I0310 18:55:36.446965 filer_grpc_server_sub_meta.go:279 + listener filer.remote.sync@127.0.0.1:50589
I0310 18:55:36.446982 filer_grpc_server_sub_meta.go:34  filer.remote.sync@127.0.0.1:50589 starts to subscribe /path/to/mount/cloud/data/directory from 2023-03-10 11:55:36.44564 +0000 GMT
```
and running the log command, I can see
```
weed -v=4 filer.remote.sync ...
I0310 18:48:58.781496 config.go:59 Reading security.toml from /etc/seaweedfs/security.toml
```

But nothing actually gets downloaded from S3 nor anything is written back.

Could you please let me know if I have missed out on any steps?

Any help is much appreciated please.

**Reply by chrislusf** [reply]
See `filer.remote.sync -h`, it only moves from SeaweedFS to S3. 

For S3 -> SeaweedFS, use `remote.cache` in `weed shell`.

**Reply by JithenJ** [reply]
@chrislusf Thanks very much for the command.
I can confirm that now filer is configured for the remote s3 storage. 

When I ran the command `weed filer.remote.sync ` it downloaded the file from remote s3 into the filer store.
And I then added a new file into the filer store by calling the API. and then when i executed `remote.cache` I could see that this file was copied to S3. 
I have a few more additional questions if I may:

**1.** My requirement is to push 100s of files into the remote repository. For my testing I was calling the endpoint
`POST /path/to/file`
to interact with the filer and add a new file, which then created an entry in the filer store and then pushed the file into s3.

But if there are a lot of files with a directory structure that I want to move over to S3.
for eg :  if I check out a repository from github into the mounted directory, how can I ensure that my entire directory structure along with all the files are copied over to S3 ?

Calling the filer endpoint for each file one by one will of course be time consuming. Surely there should be a way ? 
I looked at the Filer Java Client to see if there were methods that would create an entire directory structure with all its contents into the filer, but I could not easily find it. Can you help ?

**2.** In my setup, will the filer be storing the files both in local as well as remote ? How can I check that ? 
Ideally I only want the hot files to store in the local and if the file is warm it should just be only in the remote so that it clears up the local store for other hot files. Was there any configuration for this ?

**3.** How can I ensure that when I start up my weed s3 server, the command for `filer.remote.sync` and `remote.cache` is executed?  
The docs says I can configure cron for this ? Do you mean a normal cron that will just execute these commands at regular intervals ?

Any help is much appreciated.

**Reply by JithenJ** [reply]
@chrislusf I believe I solved the problem1. 

I ran the below command which ensured that any changes in the `/path/to/localstorage` was reflected in the filer. 
And my `/path/to/filer/remotestorage` was anyway mounted to S3, which copied it to S3. 

`weed mount -filer=localhost:8888 -dir=/path/to/localstorage -filer.path=/path/to/filer/remotestorage`

Would still appreciate if you had some suggestions for 2 and 3 please?
